src/backend/distributed/utils/colocation_utils.c

```diff
@@ -14,14 +14,322 @@
+	ErrorIfShardPlacementsNotColocated(sourceRelationId, targetRelationId);
+
+	/*
+	 * Get colocation group of the source table, if the source table does not
+	 * have a colocation group, create a new one, and set it for the source table.
+	 */
+	sourceColocationId = TableColocationId(sourceRelationId);
+	if (sourceColocationId == INVALID_COLOCATION_ID)
+	{
+		uint32 shardCount = ShardIntervalCount(sourceRelationId);
+		uint32 shardReplicationFactor = TableShardReplicationFactor(sourceRelationId);
+
+		sourceColocationId = CreateColocationGroup(shardCount, shardReplicationFactor,
+												   sourceDistributionColumnType);
+		UpdateColocationGroup(sourceRelationId, sourceColocationId);
+	}
+
+	/* finally set colocation group for the target relation */
+	UpdateColocationGroup(targetRelationId, sourceColocationId);
+
+	heap_close(pgDistColocation, NoLock);
+}
+
+
+/*
+ * ErrorIfShardPlacementsNotColocated checks if the shard placements of the
+ * given two relations are physically colocated. If they are not colocated,
+ * it errors out.
+ *
+ * Note that, this functions assumes that both tables are hash distributed.
+ */
+static void
+ErrorIfShardPlacementsNotColocated(Oid leftRelationId, Oid rightRelationId)
+{
+	List *leftShardIntervalList = NIL;
+	List *rightShardIntervalList = NIL;
+	ListCell *leftShardIntervalCell = NULL;
+	ListCell *rightShardIntervalCell = NULL;
+	char *leftRelationName = NULL;
+	char *rightRelationName = NULL;
+	uint32 leftShardCount = 0;
+	uint32 rightShardCount = 0;
+
+	/* get sorted shard interval lists for both tables */
+	leftShardIntervalList = LoadShardIntervalList(leftRelationId);
+	rightShardIntervalList = LoadShardIntervalList(rightRelationId);
+
+	/* prevent concurrent placement changes */
+	LockShardListMetadata(leftShardIntervalList, ShareLock);
+	LockShardListMetadata(rightShardIntervalList, ShareLock);
+
+	leftRelationName = get_rel_name(leftRelationId);
+	rightRelationName = get_rel_name(rightRelationId);
+
+	leftShardCount = list_length(leftShardIntervalList);
+	rightShardCount = list_length(rightShardIntervalList);
+
+	if (leftShardCount != rightShardCount)
+	{
+		ereport(ERROR, (errmsg("shard counts don't match for %s and %s",
+							   leftRelationName, rightRelationName)));
+	}
+
+	/* compare shard intervals one by one */
+	forboth(leftShardIntervalCell, leftShardIntervalList,
+			rightShardIntervalCell, rightShardIntervalList)
+	{
+		ShardInterval *leftInterval = (ShardInterval *) lfirst(leftShardIntervalCell);
+		ShardInterval *rightInterval = (ShardInterval *) lfirst(rightShardIntervalCell);
+
+		List *leftPlacementList = NIL;
+		List *rightPlacementList = NIL;
+		List *sortedLeftPlacementList = NIL;
+		List *sortedRightPlacementList = NIL;
+		ListCell *leftPlacementCell = NULL;
+		ListCell *rightPlacementCell = NULL;
+
+		uint64 leftShardId = leftInterval->shardId;
+		uint64 rightShardId = rightInterval->shardId;
+
+		bool shardsIntervalsEqual = ShardsIntervalsEqual(leftInterval, rightInterval);
```

Comment:
`ShardsIntervalsEqual ()` is too strong compared to what the function is actually doing. 
Also, we have the same function in physical planner `ShardIntervalsOverlap()`.
I think the best way to handle this is that move `ShardIntervalsOverlap()` to shardinterval_utils and use that?